server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -705,12 +705,15 @@ private boolean balanceByWeights(NodeSorter sorter) {
                             highIdx = relevantNodes - 1;
 
                             if (routingNodes.getRelocatingShardCount() > 0) {
-                                // ES-12955: Check routingNodes.getRelocatingShardCount() > 0 in case the first relocation is a THROTTLE.
-                                // This should rarely happen since in most cases, we don't throttle unless there is an existing relocation.
-                                // But it can happen in production for frozen indices when the cache is still being prepared. It can also
-                                // happen in tests because we have decider like RandomAllocationDecider that can randomly return THROTTLE
-                                // when there is no existing relocation.
                                 shardBalanced = true;
+                            } else {
+                                // A THROTTLE decision can happen when not simulating
+                                assert allocation.isSimulating() == false
+                                    : "unexpected THROTTLE decision (simulation="
+                                        + allocation.isSimulating()
+                                        + ") when balancing index ["
+                                        + index
+                                        + "]";
```

Comment:
Seeing the assert in a conditional branch makes this seem like certain paths are okay for THROTTLE. Suppose we were in a simulation and somehow got more than 1 relocating shard at once: that would be unsafe, right? Is there any reason we can't safely always have the assert in place? Like,
```
assert allocation.isSimulating() == false || routingNodes.getRelocatingShardCount() <= 1 : "....";
if (routingNodes.getRelocatingShardCount() > 0) {
    shardBalanced = true;
}
if (completeEarlyOnShardAssignmentChange && shardBalanced) {
....
```